### PR TITLE
Feature/weets api integration

### PR DESCRIPTION
**Pull request by wgermancutraro**
## Summary

Add the weets service, created a mappers folder with a .keep inside, removed the controllers folder, and refactored routes so it's all there. Also create an asyncWrapper to avoid having to put the try catch and that this is automatic via next (error).

I created a couple of tests on the service so that the hook also works when pushing (and apart because the tests are crucial!)

## Known Issues

I create a route for the weets, i know that this will be removed in a future but i need a route file to make that refactor. So you can test it going to: /api/v1/weets and with the console.

In the console.js file i remove the file find ending (...Service) because i called my file with the Service word at the end.

## Trello Card

https://trello.com/c/YDydn0YX/32-servicios-promises-witter

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -30,12 +30,17 @@ const app = express();
...
+// Swagger docs
 app.use('/docs', swaggerUi.serve, swaggerUi.setup(documentation));
 
+// logger
 if (!config.isTesting) app.use(expressMiddleware({ loggerFn: logger.info }));
 
-routes.init(app);
-
+// routes handling
+app.get('/', (_, res) => res.status(200).json(process.uptime()));
```
Comment (martinremis) [inline app.js]: Move this to a controller. Actually, it was already in a controller, why did you move it?

Comment (martinremis) [reply app.js]: In this file should not be routes handling, but routers handling

File: app.js [hunk 2]
```diff
@@ -30,12 +30,17 @@ const app = express();
...
+app.use('/api/v1', routes);
```
Comment (martinremis) [inline app.js]: We don't use routes versioning. We use versioning by header: https://github.com/Wolox/tech-guides/blob/master/backend/backend-standard-and-best-practices.md#4--api-versioning
Also, there is no need to have the `/api` sub-route, since this server will only be an api. It will not have other type of resources.

File: app/utils/asyncWrapper.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+const asyncWrapper = fn => (req, res, next, ...args) =>
```
Comment (martinremis) [inline app/utils/asyncWrapper.js:1]: I like this abstraction

File: app/services/weetsService.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const request = require('../utils/request');
+
+const weetsService = {
+  getRandomNumber() {
```
Comment (martinremis) [inline app/services/weetsService.js]: Since you do not need the context, use an arrow function here, and make it constant

File: app/services/weetsService.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    // generate a random number between 0 and 9999 (supported by numbers api)
+    return Math.floor(Math.random() * (9999 - 0 + 1) + 0);
+  },
+
+  async getOneRandomPhrase() {
```
Comment (martinremis) [inline app/services/weetsService.js]: Same, use an arrow function for this

File: app/services/weetsService.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    const { data } = await request(`/${this.getRandomNumber()}/math`);
+    return data;
+  }
+};
+
+module.exports = weetsService;
```
Comment (martinremis) [inline app/services/weetsService.js]: Here you are exporting all the methods. The idea is to export just what you need to, and use `getRandomNumber` as an internal function (kind of private)

File: test/unit/weets.service.spec.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+const weetService = require('../../app/services/weetsService');
+
+describe('Weets Service', () => {
+  it('Should be an object', () => {
+    expect(typeof weetService).toBe('object');
+  });
+
+  it('Should generate a random number between 0 and 9999', () => {
+    const number = weetService.getRandomNumber();
```
Comment (martinremis) [inline test/unit/weets.service.spec.js]: This is not part of the card, but since you are adding it, I will see it :smile: : 
We usually do unit testing, and as so, it should not call any external service. Maybe a test database (ideally not even that).
So, here you should mock the http request to avoid testing against an external service. I recommend nock library for that

Comment (martinremis) [reply test/unit/weets.service.spec.js]: There also are other things about tests , but i will save it to the testing card

File: .env.example [hunk 8]
```diff
@@ -0,0 +1,21 @@
+PORT=8081
+
+DB_HOST=localhost
+DB_PORT=5432
+DB_USERNAME=postgres
+DB_PASSWORD=Sherman33
```
Comment (martinremis) [inline .env.example]: I know it is just a local training database, but do not commit passwords :sweat_smile:

**Review by martinremis — CHANGES_REQUESTED** [review]

File: app/services/weetsService.js (hunk 4 above)
Comment (wgermancutraro) [reply app/services/weetsService.js]: i needed it because i did this.getRandomNumber but also i can do weetsService.getRandomNumber

File: app/services/weetsService.js (hunk 6 above)
Comment (wgermancutraro) [reply app/services/weetsService.js]: Perfect! But what about test that "private method"?

File: test/unit/weets.service.spec.js (hunk 7 above)
Comment (wgermancutraro) [reply test/unit/weets.service.spec.js]: I agree to mock up this test. But then in the others (for example in the tests that I am going to do for the sign up) I like to do integration testing using a build time database (mocked with jest)

I leave this arrangement for the test ticket.

File: app/services/weetsService.js (hunk 4 above)
Comment (martinremis) [reply app/services/weetsService.js]: if it is in the same scope you do not need to call it as an object method. it could be an static method in the context. 
Also, if you need to export it, then it is not a valid weetService method. In that case, it would be better to have it in a utils or helper

File: app/services/weetsService.js (hunk 6 above)
Comment (martinremis) [reply app/services/weetsService.js]: You can test it by its usage. I mean, if it is used by X method in this file, you can test that X method and the private method will be called and tested as well.

File: app/routes/index.js [hunk 9]
```diff
@@ -1,6 +1,8 @@
 const router = require('express').Router();
 const weetsRoutes = require('./weetsRoutes');
 
+router.get('/', (_, res) => res.status(200).json(process.uptime()));
```
Comment (martinremis) [inline app/routes/index.js]: Do not add controller logic in a routes file. Otherwise, you will have to test this file, and that is not nice at all :sweat_smile:.
So, It is better to have a simple controller to do the server health check

File: app/services/weetsService.js [hunk 10]
```diff
@@ -1,15 +1,14 @@
 const request = require('../utils/request');
 
 const weetsService = {
```
Comment (martinremis) [inline app/services/weetsService.js]: There is no need to have an object here. You can use `exports` to define exportable methods, or, if you want to keep using module.exports, you could just define it like:
```js
const getRandomNumber = () => Math.floor(Math.random() * (9999 - 0 + 1) + 0);
const getOneRandomPhrase = async () => {
    const { data } = await request(`/${getRandomNumber()}/math`);
    return data; 
}

module.exports = getOneRandomPhrase;
```

**Review by martinremis — CHANGES_REQUESTED** [review]

**Review by martinremis — APPROVED** [review]

**Comment by martinremis** [comment]
Well done :smile:
First PR here! :tada:
